state/watcher/txnwatcher.go

```diff
@@ -0,0 +1,317 @@
+			// so we'll retry again quickly.
+			backoff = PollStrategy.NewTimer(w.clock.Now())
+			next = w.clock.After(txnWatcherShortWait)
+		}
+	}
+}
+
+// flush sends all pending events to their respective channels.
+func (w *TxnWatcher) flush() {
+	// refreshEvents are stored newest first.
+	for i := len(w.syncEvents) - 1; i >= 0; i-- {
+		e := w.syncEvents[i]
+		w.hub.Publish(txnWatcherCollection, e)
+	}
+	w.syncEvents = w.syncEvents[:0]
+}
+
+// initLastId reads the most recent changelog document and initializes
+// lastId with it. This causes all history that precedes the creation
+// of the watcher to be ignored.
+func (w *TxnWatcher) initLastId() error {
+	var entry struct {
+		Id interface{} `bson:"_id"`
+	}
+	err := w.log.Find(nil).Sort("-$natural").One(&entry)
+	if err != nil && err != mgo.ErrNotFound {
+		return errors.Trace(err)
+	}
+	w.lastId = entry.Id
+	return nil
+}
+
+func (w *TxnWatcher) iter() mongo.Iterator {
+	return w.log.Find(nil).Batch(10).Sort("-$natural").Iter()
+}
+
+// sync updates the watcher knowledge from the database, and
+// queues events to observing channels.
+func (w *TxnWatcher) sync() (bool, error) {
+	added := false
+	// Iterate through log events in reverse insertion order (newest first).
+	iter := w.iteratorFunc()
+	seen := make(map[watchKey]bool)
+	first := true
+	lastId := w.lastId
+	var entry bson.D
+	for iter.Next(&entry) {
+		if len(entry) == 0 {
+			w.logger.Tracef("got empty changelog document")
+		}
+		id := entry[0]
+		if id.Name != "_id" {
+			w.logger.Warningf("watcher: _id field isn't first entry")
+			continue
+		}
+		if first {
+			w.lastId = id.Value
+			first = false
+		}
+		if id.Value == lastId {
+			break
+		}
+		w.logger.Tracef("got changelog document: %#v", entry)
+		for _, c := range entry[1:] {
+			// See txn's Runner.ChangeLog for the structure of log entries.
+			var d, r []interface{}
+			dr, _ := c.Value.(bson.D)
+			for _, item := range dr {
+				switch item.Name {
+				case "d":
+					d, _ = item.Value.([]interface{})
+				case "r":
+					r, _ = item.Value.([]interface{})
+				}
+			}
+			if len(d) == 0 || len(d) != len(r) {
+				w.logger.Warningf("changelog has invalid collection document: %#v", c)
+				continue
+			}
+			for i := len(d) - 1; i >= 0; i-- {
+				key := watchKey{c.Name, d[i]}
+				if seen[key] {
+					continue
+				}
+				seen[key] = true
+				revno, ok := r[i].(int64)
+				if !ok {
+					w.logger.Warningf("changelog has revno with type %T: %#v", r[i], r[i])
+					continue
+				}
+				if revno < 0 {
+					revno = -1
+				}
+				w.syncEvents = append(w.syncEvents, Change{
+					C:     c.Name,
+					Id:    d[i],
+					Revno: revno,
+				})
+				added = true
+			}
+		}
+	}
+	if err := iter.Close(); err != nil {
+		if qerr, ok := err.(*mgo.QueryError); ok {
+			// CappedPositionLost is code 136.
+			// Just in case that changes for some reason, we'll also check the error message.
+			if qerr.Code == 136 || strings.Contains(qerr.Message, "CappedPositionLost") {
```

Comment:
I don't think this is your code, but this is soo lovely...
:(

Reply:
Yeah, it is a bit ick, and to be honest, I don't think we need to care for this now.